### ISSUE TITLE
Strange behavior about yield* in iteration

### ISSUE THREAD

**Issue by Fozei** [dart-lang/sdk#49695]
In fact I am searching what's Stream behaviors like if the upStream's generating speed is faster than the down stream's consuming speed. As I know, this situation was handled in RxJava(FlatMap + BackpressureStrategy or use ConcatMap).

Soon I find strange behaviors in Dart(SDK error or miss understanding from my self ?).

I want iteration all combinations from a list, which like Pythons  `itertools.combinations([1, 2, 3, 4, 5], 2)`(not finish yet,so do not take the result seriously), in case the computing may be huge, so I use asyn*.

Below is my code, I use yield* for iteration, it perfors wrong:
```
import 'dart:io';

void main(List<String> args) {
  List<int> alist = [1, 2, 3, 4, 5];
  cNKStream(alist, 0, 2).listen((event) {
    sleep(Duration(milliseconds: 500));
    print("${event}receive---");
  }, onError: (e) {
    print(e);
  });
}

Stream<List<int>> cNKStream(List<int> elements, int from, int k) async* {
  if (elements.length < k) {
    throw Exception("n should not be smaller than k");
  }
  if (k < 1) {
    throw Exception("k should be bigger than 0");
  }
  if (from + k > elements.length) {
    return;
  }
  List<int> result = List.filled(k, 0);
  for (int i = 0; i < k; i++) {
    result[i] = elements[from + i];
  }
  print("=========send 1========");
  yield result;

  for (int i = from + k; i < elements.length; i++) {
    result[k - 1] = elements[i];
    print("=========send 2========");
    yield result;
  }
  yield* cNKStream(elements, from + 1, k);
}

```
then I change my code as below, it works:
```
import 'dart:io';

void main(List<String> args) {
  List<int> alist = [1, 2, 3, 4, 5];
  cNKStream(alist, 0, 2).listen((event) {
    sleep(Duration(milliseconds: 500));
    print("${event}receive---");
  }, onError: (e) {
    print(e);
  });
}

Stream<List<int>> cNKStream(List<int> elements, int from, int k) async* {
  if (elements.length < k) {
    throw Exception("n should not be smaller than k");
  }
  if (k < 1) {
    throw Exception("k should be bigger than 0");
  }
  while (from + k <= elements.length) {
    List<int> result = List.filled(k, 0);
    for (int i = 0; i < k; i++) {
      result[i] = elements[from + i];
    }
    print("=========send 1========");
    yield result;

    for (int i = from + k; i < elements.length; i++) {
      result[k - 1] = elements[i];
      print("=========send 2========");
      yield result;
    }
    from++;
  }
}
```
So what's the reason lead my first code performs wrong ? It’s SDK designs like this or my miss understanding about Stream ?
Thanks.

**Comment by a-siva** [comment]
@Fozei can you please provide us with the information requested above, thanks.

**Comment by Fozei** [comment]
@a-siva @lrhn Thanks for replaying and sorry for late.
I am running these code on Linux machine, the SDK version is `Dart SDK version: 2.17.6 (stable) (Tue Jul 12 12:54:37 2022 +0200) on "linux_x64"`. 

Forget about back pressure, now take `cNKStream` just a function that well generate a iterable which will send 10 times event to down stream, and delete the wrong used `sleep`.

When I first use the first method ,I want to receive ten times `send` each one followed with one time `receive` just as the second program do.

But the console like below:
```
=========send 1========
[1, 2]receive---
=========send 2========
[1, 3]receive---
=========send 2========
[1, 4]receive---
=========send 2========
[1, 5]receive---
=========send 1========
=========send 2========
[2, 4]receive---
=========send 2========
[2, 5]receive---
[2, 5]receive---
=========send 1========
=========send 2========
[3, 5]receive---
=========send 1========
[3, 5]receive---
[4, 5]receive---
```
By my initial assume，it should have the same out print as the second implementation:
```
=========send 1========
[1, 2]receive---
=========send 2========
[1, 3]receive---
=========send 2========
[1, 4]receive---
=========send 2========
[1, 5]receive---
=========send 1========
[2, 3]receive---
=========send 2========
[2, 4]receive---
=========send 2========
[2, 5]receive---
=========send 1========
[3, 4]receive---
=========send 2========
[3, 5]receive---
=========send 1========
[4, 5]receive---
```

**Comment by lrhn** [comment]
I see.

The timing of `yield* newStream()` is never going to be the same as an inline loop that directly yields the same values.

There will be extra delays, from creating the new stream and listening on it, cannot emit a value immediately because we promise not to do that in respons to a `listen` call, to waiting for the `done` event to tell that the inner stream is done.

However, the *delivery* of each event should happen synchronously, because that is necessary in order for back pressure to work correctly. 

When I'm running this with a 2.18.0 or 2.19.0 build, I see the expected interleaving, so most likely the problem was fixed as part of the 2.18 update to the `async` implementation of the VM.
From the changelog:
> Implementation of `async`/`async*`/`sync*` is revamped in Dart VM,

So, "fixed" :)

**Comment by a-siva** [comment]
@Fozei I hope you are able to upgrade to Dart 2.18 version, the update to `async` implementation in the VM was a large change and cannot be cherry picked to an over version.

Closing as fixed.
